@here/harp-mapview/lib/text/TextElementGroupState.ts

```diff
@@ -0,0 +1,150 @@
+ * @param lastFrameNumber Last frame the element was processed for rendering if available, otherwise
+ * `undefined`.
+ * @returns `undefined` if element was rejected, otherwise its current view distance.
+ */
+export type TextElementFilter = (
+    textElement: TextElement,
+    lastFrameNumber?: number
+) => number | undefined;
+
+/**
+ * `TextElementGroupState` keeps the state of a text element group while it's being rendered.
+ */
+export class TextElementGroupState {
+    // Sorted list of element states belonging to the group state. It does NOT have the same order
+    // as their elements in the group.
+    private m_textElementStates: TextElementState[];
+    private m_visited: boolean = false;
+    private m_needsSorting: boolean = false;
+
+    /**
+     * Creates the state for specified group.
+     * @param group The group of which the state will be created.
+     * @param filter Function used to do early rejection. @see [[TextElementFilter]].
+     * @param disableFading `true` if fading is disabled, `false` otherwise.
+     */
+    constructor(
+        readonly group: TextElementGroup,
+        filter: TextElementFilter,
+        disableFading: boolean
+    ) {
+        const length = group.elements.length;
+        this.m_textElementStates = new Array(length);
+        this.m_visited = true;
+
+        // TODO: HARP-7648. Reduce number of allocations here:
+        // a) Avoid creating the state for labels that don't pass early placement checks and make
+        //    this checks more strict.
+        // b) Break label state objects into a set of arrays held at group level, one for each
+        //    primitive field in the label state.
+        for (let i = 0; i < length; ++i) {
+            const textElement = group.elements[i];
+            const textDistance = filter(textElement);
+
+            const state = new TextElementState(textElement, this, textDistance, disableFading);
+            this.m_textElementStates[i] = state;
+        }
+    }
+
+    /**
+     * Indicates whether the group has been submitted to the [[TextElementsRenderer]] in the current
+     * frame.
+     */
+    get visited(): boolean {
+        return this.m_visited;
+    }
+
+    set visited(visited: boolean) {
+        this.m_visited = visited;
+    }
+
+    /**
+     * @returns True if any element visible after fading.
+     */
+    updateFading(time: number): boolean {
+        let groupVisible = false;
+        for (const elementState of this.m_textElementStates) {
+            if (elementState !== undefined) {
+                const elementVisible = elementState.updateFading(time);
+                groupVisible = groupVisible || elementVisible;
+            }
+        }
+        return groupVisible;
+    }
+
+    /**
+     * Updates the states of elements within the group.
+     * @param filter Function used to do early rejection. @see [[TextElementFilter]].
+     * @param disableFading `true` if fading is disabled, `false` otherwise.
```

Comment:
true, that's a bug. Instead of just fixing that, I'm gonna change the way we disable fading, right now it's clumsy, takes unnecessary properties in every label state and it's error prone.